Question installing cmake-built shared library to fixed destination in editable mode

Hi there, first of all thank you for all the amazing work.

I'm currently maintaining some python projects that bind to c++ code in a shared library via the python ctypes module. I have been using scikit-build with a setup.py file to build the c++ shared library and then place it relative to the python source files where the package can find it during import.

I've been working on migrating from scikit-build to scikit-build-core and have been able to get the project installing correctly by setting the output destination of the shared library file relative to `${SKBUILD_PLATLIB_DIR}`. However when the package is installed in editable mode such as during development this does not copy to shared library file to the same location relative to the python files.

Is it possible to detect in cmake when the project is built in editable mode and change the output destination accordingly? Alternatively, is there some better way for me to place and find this shared library file that I'm completely missing (tried importlib.resources but no luck)?

(privateGPT) C:\Users\AJAY\Desktop\PrivateGpt>$env:CMAKE_ARGS = "-DLLAMA_BLAS=ON -DLLAMA_BLAS_VENDOR=OpenBLAS"
The filename, directory name, or volume label syntax is incorrect.

(privateGPT) C:\Users\AJAY\Desktop\PrivateGpt>pip install llama-cpp-python
Collecting llama-cpp-python
  Using cached llama_cpp_python-0.2.20.tar.gz (8.7 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: typing-extensions>=4.5.0 in c:\users\ajay\anaconda3\envs\privategpt\lib\site-packages (from llama-cpp-python) (4.8.0)
Requirement already satisfied: numpy>=1.20.0 in c:\users\ajay\anaconda3\envs\privategpt\lib\site-packages (from llama-cpp-python) (1.26.0)
Collecting diskcache>=5.6.1 (from llama-cpp-python)
  Using cached diskcache-5.6.3-py3-none-any.whl.metadata (20 kB)
Using cached diskcache-5.6.3-py3-none-any.whl (45 kB)
Building wheels for collected packages: llama-cpp-python
  Building wheel for llama-cpp-python (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for llama-cpp-python (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      *** scikit-build-core 0.6.1 using CMake 3.27.7 (wheel)
      *** Configuring CMake...
      2023-11-30 18:40:30,114 - scikit_build_core - WARNING - Can't find a Python library, got libdir=None, ldlibrary=None, multiarch=None, masd=None
      loading initial cache file C:\Users\AJAY\AppData\Local\Temp\tmpkh6aqsh2\build\CMakeInit.txt
      -- Building for: NMake Makefiles
      CMake Error at CMakeLists.txt:3 (project):
        Running

         'nmake' '-?'

        failed with:

         The system cannot find the file specified


      CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
      CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
      -- Configuring incomplete, errors occurred!

      *** CMake configuration failed
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for llama-cpp-python
Failed to build llama-cpp-python
ERROR: Could not build wheels for llama-cpp-python, which is required to install pyproject.toml-based projects

--- a/docs/configuration/index.md
+++ b/docs/configuration/index.md
@@ -923,6 +923,8 @@ On the command line, you can pass `-Ceditable.mode=inplace` to enable this mode.
 
 :::
 
+(triggering-a-rebuild-manually)=
+
 ### Triggering a rebuild manually
 
 You don't have to enable `editable.rebuild` to rebuild on demand for the

--- a/docs/guide/cmakelists.md
+++ b/docs/guide/cmakelists.md
@@ -91,6 +91,8 @@ included in that section.
 See [search paths section](../configuration/search_paths.md) for more details on
 how the search paths are constructed.
 
+(install-directories)=
+
 ## Install directories
 
 Scikit-build-core will install directly into platlib, which will end up in

--- a/docs/guide/faqs.md
+++ b/docs/guide/faqs.md
@@ -184,9 +184,9 @@ In redirect-mode editable installs (the default), `importlib.resources` finds
 the compiled library through the redirecting finder, so the code above works
 unchanged. Note that accessing a resource does **not** trigger a rebuild — plain
 libraries are not importable modules, so the automatic `editable.rebuild`
-on-import hook does not fire for them. To pick up C/C++ changes, either request a
-rebuild explicitly (this works whenever a persistent `build-dir` is set, with or
-without `editable.rebuild`)…
+on-import hook does not fire for them. To pick up C/C++ changes, either request
+a rebuild explicitly (this works whenever a persistent `build-dir` is set, with
+or without `editable.rebuild`)…
 
 ```python
 import mypackage

--- a/docs/guide/faqs.md
+++ b/docs/guide/faqs.md
@@ -118,6 +118,94 @@ wheel-repair tool. See [](#dynamic-linking) for all of the options.
 
 [GNUInstallDirs]: inv:cmake:cmake:module#module:GNUInstallDirs
 
+## Shipping a library to load with `ctypes`
+
+If your package is a thin `ctypes` (or `cffi`) wrapper around a CMake-built
+shared library, rather than a compiled Python extension, the goal is to install
+the library alongside your Python code and then locate it at runtime without
+hard-coding a path. This has the nice property that a single wheel works on
+every Python version, since the library does not touch the Python ABI. The
+tradeoff is that you have to find and load the library yourself.
+
+### Install the library next to your Python code
+
+Point the install destination at your package directory so the library lands in
+`site-packages/mypackage/` next to `__init__.py`:
+
+```cmake
+install(TARGETS mylib DESTINATION mypackage)
+```
+
+The destination is relative to the platlib (`${SKBUILD_PLATLIB_DIR}`) by
+default; you can name any of the [install trees](#install-directories)
+explicitly if you need to. If you set `wheel.install-dir = "mypackage"`, then
+the destination is relative to that instead, and a bare `DESTINATION .` works.
+
+### Find it at runtime with `importlib.resources`
+
+Do **not** compute the path relative to `__file__` — that assumes the package
+lives on a real filesystem, which is not guaranteed (it could be in a zip, and
+in an editable install the Python source and the compiled library live in
+different directories). Use `importlib.resources` instead, which
+scikit-build-core's editable installs fully support:
+
+```python
+import ctypes
+import sys
+from importlib.resources import files
+
+# Pick the right suffix for the platform.
+_suffix = {"win32": ".dll", "darwin": ".dylib"}.get(sys.platform, ".so")
+_lib = files("mypackage") / f"libmylib{_suffix}"
+
+lib = ctypes.CDLL(str(_lib))
+```
+
+For the general (zip-safe) case, wrap the traversable in
+`importlib.resources.as_file`, which extracts the resource to a real path if
+necessary. Because `ctypes` needs the file to remain on disk for the lifetime of
+the process, keep the context manager open — for example with an
+`contextlib.ExitStack` closed at interpreter exit:
+
+```python
+import atexit, ctypes
+from contextlib import ExitStack
+from importlib.resources import files, as_file
+
+_files = ExitStack()
+atexit.register(_files.close)
+_lib = _files.enter_context(as_file(files("mypackage") / f"libmylib{_suffix}"))
+lib = ctypes.CDLL(str(_lib))
+```
+
+### Editable installs and rebuilds
+
+In redirect-mode editable installs (the default), `importlib.resources` finds
+the compiled library through the redirecting finder, so the code above works
+unchanged. Note that accessing a resource does **not** trigger a rebuild — plain
+libraries are not importable modules, so the automatic `editable.rebuild`
+on-import hook does not fire for them. To pick up C/C++ changes, either request a
+rebuild explicitly (this works whenever a persistent `build-dir` is set, with or
+without `editable.rebuild`)…
+
+```python
+import mypackage
+
+mypackage.__loader__.rebuild()
+```
+
+…or import a real extension module from the same project first, which does fire
+the on-import hook when `editable.rebuild` is enabled. See
+[](#triggering-a-rebuild-manually) for the details and the `build-dir`
+requirement.
+
+### Runtime search paths
+
+If your shipped library links against _other_ shared libraries, you still need
+to make those discoverable at load time (`RPATH` on Linux/macOS,
+`os.add_dll_directory` on Windows). See [](#dynamic-linking) for the full set of
+options, including wheel-repair tools.
+
 ## Repairing wheels
 
 Like most other backends[^1], scikit-build-core produces `linux` wheels, which

